### ISSUE TITLE
Is this repository supposed to be empty?

### ISSUE THREAD

**Issue by nwidger** [caddy-dns/linode#1]
I stumbled upon this repo while trying to figure out how to enable wildcard certs in Caddy via Let's Encrypt's DNS challenge:

https://caddyserver.com/docs/automatic-https#dns-challenge

which led me to this wiki article:

https://caddy.community/t/how-to-use-dns-provider-modules-in-caddy-2/8148

which led me to this repository, but, curiously, it's empty.  However, I then searched for `linode` on https://caddyserver.com/download and found this plugin:

https://github.com/tosie/caddy-dns-linode

which isn't empty and looks like it's what I need but is potentially less official?  Anyways, I'm just curious if this situation is expected or if I'm missing something.  Is https://github.com/tosie/caddy-dns-linode the best plugin to use if I'm using Linode as my DNS provider currently?  My plan is to build a custom Caddy binary with `xcaddy` and am trying to figure out which `--with` flag to use.  Unsurprisingly, using `github.com/caddy-dns/linode` fails:

```
$ xcaddy build --output ./caddy --with github.com/caddy-dns/linode
< ... snip ... >
2023/06/21 16:16:01 [INFO] exec (timeout=-2562047h47m16.854775808s): /usr/devel/bin/go get -d -v github.com/caddy-dns/linode github.com/caddyserver/caddy/v2 
go: github.com/caddy-dns/linode: no matching versions for query "upgrade"
2023/06/21 16:16:02 [FATAL] exit status 1
```

where as `github.com/tosie/caddy-dns-linode` works:

```
$ xcaddy build --output ./caddy --with github.com/tosie/caddy-dns-linode
< ... snip ... >
2023/06/21 16:16:30 [INFO] exec (timeout=-2562047h47m16.854775808s): /usr/devel/bin/go build -o /home/niels/caddy -ldflags -w -s -trimpath 
2023/06/21 16:16:31 [INFO] Build complete: ./caddy
2023/06/21 16:16:31 [INFO] Cleaning up temporary folder: /tmp/buildenv_2023-06-21-1616.2735815113
$ ./caddy list-modules 
< ... snip ... >
tls.stek.distributed
tls.stek.standard

  Standard modules: 100

dns.providers.linode

  Non-standard modules: 1

  Unknown modules: 0
```

And thanks for Caddy, it's amazing!  I'm also happy to open a PR somewhere if that would help.

**Comment by mholt** [comment]
Thanks for the feedback! And good question. I hadn't realized this repo is still empty.

When someone contributes a [libdns package](https://github.com/libdns/libdns) I also create a repo in this org because the vast majority of libdns contributors are implementing for use in Caddy. So usually they spend a few minutes to create a caddy-dns package as well. (There's a template and it's mostly copy-paste, change just a few things, and you're done.)

I wasn't aware of that other package, it might be for an older version of the Linode API. :man_shrugging: I'd rather have a package here in this org though.

@bcrochet Thanks for implementing libdns/linode! Are you a Caddy user, and if so, do you have a caddy-dns package that you forgot to push, maybe? If not, no worries. We can let anyone maintain this repo who has a need for it.

@nwidger We can also add you to this repo if you'd want to contribute.

**Comment by nwidger** [comment]
@mholt Thanks for the quick response!  I'd be happy to be added to this repo and add an initial implementation, I looked at the template and it does indeed look pretty simple.  I'm also fine if @bcrochet wants to do it instead, or if he'd like to take over this repo at some point later down the road.  I just want a Linode Caddy plugin that is officially supported.  A few questions:

- https://github.com/tosie/caddy-dns-linode already exists and is discoverable via https://caddyserver.com/download.  Is it a problem if both it and https://github.com/caddy-dns/linode use the `dns.providers.linode` module name?  I haven't looked at the code, but I'd assume this would be some kind of run-time error if someone actually tried to build a `caddy` binary containing both plugins.  I'm going to assume it's fine for this repo to also name its plugin `dns.providers.linode`, but I figured I'd ask anyways.

- It looks like https://github.com/libdns/linode doesn't have any tags, I'm assuming this repo would prefer to refer to an actual release in its `go.mod` file.  @bcrochet Would it be possible to add something like a `v0.1.0` tag to the repo, please?

**Comment by nwidger** [comment]
One additional question:

- Am I right in assuming that both the `APIToken` and `Domain` fields in `linode.Provider` would need to be filled out for the plugin to function correctly (https://github.com/libdns/linode/blob/main/provider.go#L15)?  I.e. the Caddy plugin would need both `api_token` and `domain` (and not just `api_token`) parameters in the Caddyfile.

**Comment by mholt** [comment]
> . I just want a Linode Caddy plugin that is officially supported.

Just to clarify, none of the DNS provider plugins are "official" or supported by the Caddy team (we don't use all these DNS providers so we can't support them). However I can probably help maintain some with a sponsorship sufficient to fund that.

> Am I right in assuming that both the APIToken and Domain fields in linode.Provider would need to be filled out for the plugin to function correctly (https://github.com/libdns/linode/blob/main/provider.go#L15)? I.e. the Caddy plugin would need both api_token and domain (and not just api_token) parameters in the Caddyfile.

That, I am not sure -- @bcrochet would have to answer that.

It does seem odd if the config is limited to a single domain.

**Comment by nwidger** [comment]
> Just to clarify, none of the DNS provider plugins are "official" or supported by the Caddy team (we don't use all these DNS providers so we can't support them). However I can probably help maintain some with a sponsorship sufficient to fund that.

That makes perfect sense, and I didn't mean to imply I was owed anything or anything along those lines.  I just want the next person who follows the guide in the wiki article to find a Linode plugin and not scratch their head in confusion. :smiley:

**Comment by nwidger** [comment]
@mholt I followed the template and believe I have an initial attempt at a Caddy Linode plugin.  However, you can't fork or open a PR on an empty repo. :disappointed: If you want to add me to this repo, I'd be happy to either commit what I have to `main` or open a new PR targeting `main`, otherwise we can wait and see what @bcrochet wants to do.

**Comment by mholt** [comment]
I see no reason to wait, a contribution is a contribution :)

Invited you to be a maintainer. Thank you!

**Comment by nwidger** [comment]
@mholt @bcrochet I've pushed an initial implementation of the Caddy Linode plugin to this repo, but I think we need to make some changes to https://github.com/libdns/linode before it's actually usable.  I've opened a PR https://github.com/libdns/linode/pull/1 in that repo with what I think are reasonable changes.  I don't think previously the `APIToken` field in the `Provider` struct was ever actually used to configure the `linodego.Client` (meaning any requests sent by it would not have been authenticated), therefore I updated `Provider.init` to set the token on the client.  I also added `api_url` and `api_version` fields in an attempt to at least offer the same level of customization as https://pkg.go.dev/github.com/linode/linodego#NewClientFromEnv.  If everyone is okay with those changes, once they get merged it would be great to tag https://github.com/libdns/linode with a `v0.1.0` release and then I will update this repo to refer to the tag and support using the new `api_url`/`api_version` fields.  Once that is all taken care of I think we'll be ready to close this issue.  How does that all sound?

**Comment by nwidger** [comment]
@mholt @bcrochet Okay, I think everything is in a good place now.  https://github.com/libdns/linode/pull/1 is merged and tagged with a v0.2.0 release, and this repo has been updated to use the new tag and has its own v0.2.0 tag.  I also registered the new `dns.providers.linode` module so it shows up on https://caddyserver.com/download and https://caddyserver.com/docs/modules/dns.providers.linode.

Thanks @mholt for your help with all this!  Let me know if there's anything I missed.
